File: handler.pl
#!/usr/bin/env perl

#
# Handler for running a
#

use strict;
use warnings;

use CGI qw(-debug);
use Digest::SHA qw(hmac_sha256_hex);
use feature qw(say);

print "Content-type: text/plain\n\n";

my $secret = $ENV{"SECRET"};
if (not defined $secret) {
    die "No secret set in environment! Aborting...\n";
}

my $sync_dir = $ENV{"SYNC_DIR"};
if (not defined $sync_dir) {
    die "No sync directory in environment (SYNC_DIR)! Aborting...\n";
}
elsif (!-e $sync_dir) {
    die "Invalid sync directory '$sync_dir'. Aborting...\n";
}

my $q = CGI->new;
if (my $data = $q->param('POSTDATA')) {

    my $check = $ENV{"HTTP_X_HUB_SIGNATURE_256"};
    my $signature = "sha256=" . hmac_sha256_hex($data, $secret);

    if ($signature ne $check) {
        die "Mismatching secret key! Aborting...\n";
    }

    chdir $sync_dir;
    system("git", "pull");
    my @files = glob("*.gpkg");
    my @cmd = ("geoserver-sync", "--user", "admin", "--workspace", "ehri", @files);
    say join(" ", @cmd);
    if (system(@cmd)) {
        die "Sync script exited with non-zero code: $!\n";
    }
}

File: handler.py
#!/usr/bin/env python3

import glob
import hashlib
import hmac
import os
import sys


secret = os.environ.get("SECRET", "")
sync_dir = os.environ.get("SYNC_DIR")
if sync_dir is None or not os.path.isdir(sync_dir):
    raise Exception("Missing or invalid SYNC_DIR in environment")

data = sys.stdin.read()
check = os.environ.get("HTTP_X_HUB_SIGNATURE_256")
signature = "sha256=" + hmac.new(str.encode(secret), data.encode('utf-8'), digestmod=hashlib.sha256).hexdigest()
if signature != check:
    print("Content-type: text/plain")
    print("Status: 400 Bad Request", end="\r\n\r\n", flush=True)
    print("Unexpected payload")
else:
    print("Content-type: text/plain", end="\r\n\r\n", flush=True)
    os.chdir(sync_dir)
    os.system("git pull")
    files = " ".join(glob.glob("*.gpkg"))
    cmd = "geoserver-sync --user admin --workspace ehri " + files
    if os.system(cmd) != 0:
        raise Exception(f"Command '{cmd}' exited with non-zero code")